src/posts/2022-11-27-define-once/define-once.md

```diff
@@ -0,0 +1,355 @@
+
+```python
+@cache
+def long_running_notebook_function(*args, **kwargs):
+    # Now our notebook will no longer take forever to run!
+    return result
+```
+
+We use this new function with reckless abandon in our code, and it significantly
+reduce the time it takes a cell in the notebook to run. At some point we want to
+rerun everything in the notebook (to say regenerate some plots with new
+`matplotlib.rcParams` settings), and we use the "Run All Cells" menu item
+
+...
+
+and it now takes _forever_ to run the same cells that. What happened?!
+
+
+## Functions are redefined every time a cell is run
+
+When using a notebook, the code in a cell is rerun every time we execute that
+cell. The definitions in the cell being run are assigned to the global namespace
+in the running notebook kernel. Put another way, we can think of running a cell
+as the following pseudocode.
+
+```python
+# Before running a cell, we can find find all globally defined
+# variables in the `globals()` dictionary. This will include
+# a pointer to all of our defined functions, plus a bunch of
+# stuff defined by the Python kernel/Jupyter notebook internals.
+before = globals()
+
+# Take the source code in the cell as a string called `cell_str`.
+# Anything defined in the cell string will be merged into the 
+# `globals()` dictionary.
+exec(cell_str)
+
+# The globals dictionary should now be modified with whatever was
+# defined within `cell_str`.
+after = globals()
+
+# The following equation relates the two dictionaries
+# (minus some jupyter bookkeeping)
+#
+#     after = before | top_level_definitions(cell_str)
+#
+# where `|` is the dictionary merge operator from Python 3.9 and
+# `top_level_defintions` returns the top level definitions in the
+# string along with the associated function pointers created by `exec`.
+```
+
+The reason the `cache` decorator does not seem to work is that when we run a
+cell, a _new function with the same name_ gets created and replaces the existing
+function in the global namespace. We lose track of our first cached function!
+
+
+## But all is not lost!
+
+For the `cache` decorator to working even after running the cell that defines
+the desired function multiple times, we need to keep track of the original
+function with the cache attached and ignore the new function created when the
+cell is run again. Conveniently, the `globals()` dictionary gives us access to
+the global namespace, enabling us to see if a function with a given name already
+exists and hold onto the original definition instead of the new (likely
+equivalent) definition.  The name of a function can be looked up through the
+`__name__` property, so we can define the following decorator[^2] to find the
+existing function and keep it.
+
+[^2]: A decorator is just a higher order function (a function that takes another
+      function as an argument).
+
+```python
+def define_once_first_try(f):
+    if f.__name__ in globals().keys():
```

Comment:
Interesting! So when you call Run All Cells, the global namespace isn't cleared?

Reply:
Nope! Just added to. There is a "Restart and Run All Cells" option, which will clear the namespace.